# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студенты гр. 2384 Пчелинцева Кристина и Лукьянова Мария.

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Основные теоретические положения

Если случайны величины независимы, то $f(x,y)=f_1(x)\cdot f_2(y)$ иначе между ними может быть:

1) функциональная зависимость:
$y = g(x)$

2) статистическая зависимость:
$\phi (x|y) = f(x,y)/f_2(y)$

$\phi (y|x) = f(x,y)/f_1(x)$

Корелляционная зависимость - статистичеакая зависимость.

**Коэффициент корелляции**
$$r_{xy} = \frac{\mu_{xy}}{\sigma_x \sigma_y},$$
для него справедливо: $|r_{xy}|\le 1$'

Справедливы утверждения:

X и Y кореллированы -> СВ X и Y зависимы

X и Y независимы -> СВ X и Y некореллированы




## Постановка задачи

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин $X$ (первый признак) и $Y$ (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

## Пункт 1.
Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации). Оформить результаты в виде таблицы, сделать выводы.

Данные рассчеты уже были проведены и проанализированы в соответсвующих лабораторных работах. Отобразим результаты в виде таблицы.

|$\bar X$ |$\mathbb{D_в}$| $\sigma_в$  |$\bar A_s$ | $\bar E$ | $M_o^*$ |$M_e^*$ | $V^*$ |
|---|---|---|---|---|---|---|---|
|  3016.5851 |  636182.5062 | 797.6105  |  1.0357 |  0.67 |  2646.4726 | 2838.8412  |  26% |



### Пункт 2.
Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.

In [112]:
import numpy as np
import pandas as pd

In [113]:
df = pd.read_csv("team33_pr1.csv")
data = np.array(df)

In [114]:
k = 8

bins1 = pd.cut(df["PS"], bins=k, right=False, precision=6)
bins2 = pd.cut(df["PPY"], bins=k, right=False, precision=6)

contingency_table = pd.crosstab(bins2, bins1, dropna=False)

print(contingency_table.to_markdown())

| PPY                        |   [84.812623, 92.711045) |   [92.711045, 100.609467) |   [100.609467, 108.50789) |   [108.50789, 116.406312) |   [116.406312, 124.304734) |   [124.304734, 132.203156) |   [132.203156, 140.101578) |   [140.101578, 148.063187) |
|:---------------------------|-------------------------:|--------------------------:|--------------------------:|--------------------------:|---------------------------:|---------------------------:|---------------------------:|---------------------------:|
| [1749.166667, 2235.208333) |                        0 |                         7 |                         1 |                         3 |                          2 |                          0 |                          0 |                          0 |
| [2235.208333, 2721.25)     |                        0 |                         6 |                         2 |                        15 |                         12 |                          0 |                          0

### Пункт 3.
По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.

In [115]:
correlation_table = pd.crosstab(bins2, bins1, margins=True, margins_name="Σ")

correlation_table.rename(index={"Σ": "$n_y$"}, columns={"Σ": "$n_x$"}, inplace=True)

print(correlation_table.to_markdown())

| PPY                        |   [84.812623, 92.711045) |   [92.711045, 100.609467) |   [100.609467, 108.50789) |   [108.50789, 116.406312) |   [116.406312, 124.304734) |   [132.203156, 140.101578) |   [140.101578, 148.063187) |   $n_x$ |
|:---------------------------|-------------------------:|--------------------------:|--------------------------:|--------------------------:|---------------------------:|---------------------------:|---------------------------:|--------:|
| [1749.166667, 2235.208333) |                        0 |                         7 |                         1 |                         3 |                          2 |                          0 |                          0 |      13 |
| [2235.208333, 2721.25)     |                        0 |                         6 |                         2 |                        15 |                         12 |                          0 |                          0 |      35 |
| [2721.25, 3207.291667)     |              

Корреляционная таблица построена корректно (суммы частот совпадают). Наибольшее количество автомобилей (около 54%) имеют мощность 108–124 л.с. и цену 2235–3207 фунтов (около 60% выборки). Прослеживается тенденция к положительной связи между мощностью двигателя и ценой за год, однако для подтверждения этого вывода необходим расчет коэффициента корреляции.

### Пункт 4.
Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.

$\bar r_{xy} = \frac{\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij}y_ix_j - N\bar x_в \bar y_в}{NS_xS_y}$

In [116]:
intervals1 = bins1.cat.categories
intervals2 = bins2.cat.categories

n_x = np.array(contingency_table.sum(axis=0))
n_y = np.array(contingency_table.sum(axis=1))

x = np.array([(interval.left + interval.right) / 2 for interval in intervals1])
y = np.array([(interval.left + interval.right) / 2 for interval in intervals2])

cont_matrix = np.array(contingency_table)
N = np.sum(cont_matrix)


x_mean = np.sum(n_x * x) / N
y_mean = np.sum(n_y * y) / N

D_x = np.sum(n_x * (x - x_mean)**2) / N
D_y = np.sum(n_y * (y - y_mean)**2) / N

S_x = np.sqrt(N / (N - 1) * D_x)
S_y = np.sqrt(N / (N - 1) * D_y)

sum_xy = 0
for i in range(len(y)):     
    for j in range(len(x)): 
        sum_xy += cont_matrix[i, j] * x[j] * y[i]

cov = (sum_xy - N * x_mean * y_mean) / N
r_xy = cov / (S_x * S_y)

print(f"r_xy = {r_xy:.6f}")

r_xy = 0.329200


In [117]:
C_x = x[np.argmax(n_x)]
C_y = y[np.argmax(n_y)]

h_x = intervals1[0].right - intervals1[0].left
h_y = intervals2[0].right - intervals2[0].left

u = (x - C_x) / h_x
v = (y - C_y) / h_y

u_mean = np.sum(n_x * u) / N
v_mean = np.sum(n_y * v) / N

u2_mean = np.sum(n_x * u**2) / N
v2_mean = np.sum(n_y * v**2) / N

sigma_u = np.sqrt(u2_mean - u_mean**2)
sigma_v = np.sqrt(v2_mean - v_mean**2)

sum_uv = 0
for i in range(len(v)):      
    for j in range(len(u)):  
        sum_uv += cont_matrix[i, j] * u[j] * v[i]

cov_uv = (sum_uv - N * u_mean * v_mean) / N
r_xy = cov_uv / (sigma_u * sigma_v)

print(f"r_xy = {r_xy:.6f}")

r_xy = 0.332193


Значение выборочного коэффициента корреляции было вычислено двумя способами: 
по стандартной формуле и методом условных вариант. 
Получены следующие результаты:


$r_{xy}^{\text{станд}} = 0.329, \quad r_{xy}^{\text{усл}} = 0.332.$

Небольшое расхождение ($\Delta r = 0.003$) обусловлено погрешностями 
округления при вычислении условных вариант и использованием приближённых 
значений длин интервалов. В пределах точности вычислений результаты 
можно считать совпадающими, что подтверждает корректность применения 
метода условных вариант.

Полученное значение $r_{xy} = 0.33 $ указывает на наличие 
умеренной положительной связи между мощностью двигателя 
PS и ценой за год PPY. Это означает, что с ростом 
мощности двигателя цена за год имеет тенденцию к увеличению, 
однако связь не является сильной.

### Пункт 5.
Построить доверительный интервал для коэффициента корреляции при уровне значимости
γ
∈
{
0.95
,
0.99
}
, сделать выводы.

In [118]:
import scipy.stats as stats

def confidence_interval_r(r_xy, N, gamma):
    z = 0.5 * np.log((1 + r_xy)/(1 - r_xy))
    alpha = 1 - gamma
    _lambda = stats.norm.ppf(1 - alpha / 2)

    left  = z - _lambda / np.sqrt(N - 3)
    right = z + _lambda / np.sqrt(N - 3)

    left  = (np.exp(2 * left) - 1) / (np.exp(2 * left) + 1)
    right = (np.exp(2 * right) - 1) / (np.exp(2 * right) + 1)

    return left, right

left_95, right_95 = confidence_interval_r(r_xy, N, 0.95)
left_99, right_99 = confidence_interval_r(r_xy, N, 0.99)

print(f"Доверитиельный интервал r: ({left_95:.4f}; {right_95:.4f})")
print(f"Доверитиельный интервал r: ({left_99:.4f}; {right_99:.4f})")

Доверитиельный интервал r: (0.1554; 0.4883)
Доверитиельный интервал r: (0.0971; 0.5322)


С вероятностью 95% истинное значение коэффициента корреляции между мощностью двигателя (PS) и ценой за год (PPY) находится в интервале от 0.152 до 0.486. С вероятностью 99% - в интервале от 0.094 до 0.530. Поскольку оба интервала не содержат нуля, можно утверждать, что связь между признаками является статистически значимой. Положительные значения интервалов подтверждают наличие прямой корреляционной связи.

### Пункт 6.

##### Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости α = 0.05, сделать выводы.

In [119]:
# проверка гипотезы
alpha = 0.05
t_sample = r_xy * np.sqrt(N-2) / np.sqrt(1 - r_xy**2)
t_crit = stats.t.ppf(1-alpha, N-2)

print(f't выборочное  = {t_sample:.4f}')
print(f't критическое = {t_crit:.4f}')
if t_sample <= t_crit:
    print('H0 не отвергается')
else:
    print('Отвергаем гипотезу H0')

t выборочное  = 3.6770
t критическое = 1.6590
Отвергаем гипотезу H0


Поскольку $t_{выб} > t_{крит}$ , есть основания отвергать нулевую гипотезу.Таким образом, коэффициент корреляции статистически значимо отличается от нуля, что подтверждает наличие корреляционной связи между мощностью двигателя (PS) и ценой за год (PPY).

## Выводы

В ходе выполнения практической работы была проведена статистическая обработка второй выборки (цена за год, PPY) с использованием программных средств, разработанных в рамках предыдущих практических работ (№1–2). Это позволило собрать необходимые данные для исследования корреляционной зависимости между мощностью двигателя (PS) и ценой за год (PPY).

Для анализа взаимосвязи между величинами была сформирована двумерная выборка объёмом N = 111 и построена корреляционная таблица. На основе полученных данных был рассчитан коэффициент корреляции Пирсона двумя способами: по стандартной формуле и методом условных вариант. Получены следующие результаты:

$r_{xy}^{\text{станд}} = 0.329,\quad r_{xy}^{\text{усл}} = 0.332.$

Небольшое расхождение ($\Delta r = 0.003$) обусловлено погрешностями округления при вычислении условных вариант и использованием приближённых значений длин интервалов. В пределах точности вычислений результаты можно считать совпадающими, что подтверждает корректность применения метода условных вариант. Полученное значение ($r_{xy} = 0.33$) указывает на наличие умеренной положительной связи между исследуемыми признаками.

Для оценки значимости коэффициента корреляции были построены доверительные интервалы:

$\gamma = 0.95:\; (0.152;\; 0.486),\quad \gamma = 0.99:\; (0.094;\; 0.530).$

Поскольку оба интервала не содержат нуля, можно сделать вывод, что корреляционная зависимость между мощностью двигателя и ценой за год является статистически значимой.

Дополнительно была проверена гипотеза о равенстве коэффициента корреляции нулю $H_0: r_{xy} = 0$ при уровне значимости $\alpha = 0.05$. Для этого были вычислены выборочное значение статистики критерия и критическое значение. Так как $t_{\text{набл}} > t_{\text{крит}}$, нулевая гипотеза отвергается. Это подтверждает, что коэффициент корреляции статистически значимо отличается от нуля.

Таким образом, проведённое исследование показало наличие умеренной положительной корреляционной зависимости между мощностью двигателя (PS) и ценой за год (PPY). С ростом мощности двигателя цена за год имеет тенденцию к увеличению.